File: 1360-maximum-length-of-a-concatenated-string-with-unique-characters/maximum-length-of-a-concatenated-string-with-unique-characters.php
class Solution {
    private $a;

    public function __construct() {
        $this->a = ord('a');
    }

    function mask($str) {
        $mask = 0;
        for ($i = strlen($str) - 1; $i >= 0; $i--) {
            $bit = 2 ** (ord($str[$i]) - $this->a);
            if ($mask & $bit) {
                return -1;
            }
            $mask |= $bit;
        }
        return $mask;
    }

    /**
     * @param String[] $arr
     * @return Integer
     */
    function maxLength($arr) {
        $masks = [0];
        foreach ($arr as $s) {
            $sMask = $this->mask($s);
            if ($sMask == -1) {
                continue;
            }
            foreach ($masks as $mask) {
                if ($mask & $sMask) {
                    continue;
                }
                $masks[] = $mask | $sMask;
            }
        }

        $max = 0;
        foreach ($masks as $mask) {
            $max = max($max, substr_count(decbin($mask), '1'));
        }
        return $max;
    }
}

File: 1360-maximum-length-of-a-concatenated-string-with-unique-characters/maximum-length-of-a-concatenated-string-with-unique-characters.cs
public class Solution {
    public int MaxLength(IList<string> arr) {
        List<int> uniqueMasks = new List<int> { 0 };
        int maxLength = 0;

        foreach (string word in arr) {
            int mask = 0;
            bool isUnique = true;

            foreach (char c in word) {
                int charMask = 1 << (c - 'a');
                if ((mask & charMask) > 0) {
                    isUnique = false;
                    break;
                }
                mask |= charMask;
            }

            if (isUnique) {
                List<int> newMasks = new List<int>();
                foreach (int existingMask in uniqueMasks) {
                    if ((existingMask & mask) == 0) {
                        int combinedMask = existingMask | mask;
                        newMasks.Add(combinedMask);
                        maxLength = Math.Max(maxLength, CountSetBits(combinedMask));
                    }
                }
                uniqueMasks.AddRange(newMasks);
            }
        }

        return maxLength;
    }

    private int CountSetBits(int n) {
        int count = 0;
        while (n > 0) {
            count += n & 1;
            n >>= 1;
        }
        return count;
    }
}